File: dynamic_programming/198_house_robber.cc
#include <iostream>
#include <vector>

// 2024/01/12: 不看答案自己实现.

int Rob(std::vector<int>& nums) {
  if (nums.size() == 1) {
    return nums[0];
  }
  int n = nums.size();
  int pre_pre = nums[0];
  int pre = std::max(nums[0], nums[1]);
  for (int i = 2; i < nums.size(); ++i) {
    int cur = std::max(pre, pre_pre + nums[i]);
    pre_pre = pre;
    pre = cur;
  }
  return pre;
}

int main(int argc, char* argv[]) {
  std::vector<int> nums = {1, 2, 3, 1};
  std::cout << Rob(nums) << std::endl;
}

File: dynamic_programming/198_house_robber.py
from typing import List


def rob(nums: List[int]) -> int:
    dp = [0] * len(nums)

    for i in range(len(nums)):
        if i == 0:
            dp[i] = nums[i]
            continue
        if i == 1:
            dp[i] = max(nums[i], nums[i - 1])
            continue
        dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])

    return dp[-1]


nums1 = [1, 2, 3, 1]
nums2 = [2, 7, 9, 3, 1]
nums3 = [0]
print(rob(nums3))
